auto-patchelf: refactor structuredAttrs support

wolfgangwalther opened the pull request:
## Description of changes

Follow up to #318614 to use `concatTo` in `setup-hooks/auto-patchelf.sh`.

This needs to support `*` in the non-structuredAttrs case, so adjusted `concatTo` for that.

Can be tested via `nix-build -A tests.auto-patchelf-hook`. This will fail - but in the same way as on current master and the output clearly shows that flags are passed correctly. Not sure whether it is *supposed* to fail, though? CC @ConnorBaker and @yannham, who introduced it in #272752.

CC @emilazy @philiptaron @tie

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wolfgangwalther committed:
auto-patchelf: refactor structuredAttrs support

stdenv now provides better tooling to support structuredAttrs without
depending on $__structuredAttrs itself.

tie commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -409,8 +409,10 @@ concatTo() {
+                        set -o noglob
                         # shellcheck disable=SC2206
                         targetref+=( ${nameref-} )
+                        set +o noglob
```
You can use `local -` to make shell options local to the function call. Technically, this requires Bash 4.4, but see #340765.

https://www.gnu.org/software/bash/manual/html_node/Bash-Builtins.html#index-local:~:text=If%20name%20is%20,local%20is%20invoked

tie reviewed (approved):
LGTM (with a single suggestion comment).

ConnorBaker commented:
Hello! I'll take a look at this when I'm able, but don't let me approval be a blocker.

I've noticed a lot of PRs targeting staging recently dealing with various aspects of support for `__structuredAttrs` -- is there a tracking issue for better support I can use to look at recent changes? For example, I'm not familiar with the function being used in this PR, `concatTo`.

ConnorBaker commented:
> > I've noticed a lot of PRs targeting staging recently dealing with various aspects of support for `__structuredAttrs` -- is there a tracking issue for better support I can use to look at recent changes? For example, I'm not familiar with the function being used in this PR, `concatTo`.
> 
> There is no tracking issue, the discussion around this all happened in #318614.

I updated an old tracking issue: https://github.com/NixOS/nixpkgs/issues/205690. If you do additional PRs, would you mind referencing it? Makes it easier to track related work (which will come in handy when I eventually need to update or troubleshoot the setup hooks the CUDA ecosystem uses).

yannham commented:
I haven't looked at it yet, but will try to do it by tomorrow. I originally wrote the tests for #272752 - I don't think they were supposed to fail, but I can't be 100% sure either honestly as it's been some time. Will at least take a quick look at the output of `nix-build -A tests.auto-patchelf-hook`

yannham commented:
@wolfgangwalther Thanks for pinging me. The test is broken on master, and fails at `[auto-patchelf-hook-test]: Check that the runpath has the right number of entries`, which differs between the actual number of entries in the RUNPATH and the expected number of entries.

I suspect it's a change in stdenv or something, but this simple patch should fix the test (at least it does on master and it's unrelated to what we're interested in testing):

```diff
--- a/pkgs/test/auto-patchelf-hook/package.nix
+++ b/pkgs/test/auto-patchelf-hook/package.nix
@@ -60,7 +60,6 @@ stdenv.mkDerivation {
   # Additional phase performing the actual test.
   installCheckPhase =
     let allDeps = runtimeDependencies ++ [
-        (lib.getLib stdenv.cc.libc)
         (lib.getLib freetype)
       ];
     in
```

If you want to make sure your change didn't break the test, given the one line diff, you can include  the fix in this PR I guess.

wolfgangwalther committed:
auto-patchelf: fix test for hook

Fix suggested by Yann Hamdaoui in
https://github.com/NixOS/nixpkgs/pull/340858#issuecomment-2344155304.

philiptaron reviewed (approved):
I did a few builds and everything came up roses. Let's merge this in the next day or so.

jcollie commented:
The Zig build hook appears to have been broken by these changes.

philiptaron commented:
> The Zig build hook appears to have been broken by these changes.

Do you mean the autoPatchelfHook ones specifically, or a previous PR @jcollie?

jcollie commented:
> > The Zig build hook appears to have been broken by these changes.
> 
> Do you mean the autoPatchelfHook ones specifically, or a previous PR @jcollie?

I'm not exactly sure which PR broke what, but I see this now when trying to build a Zig package:

```/nix/store/3bwyb6sw8qzir9prnh3609anzw9nvp6g-zig-hook/nix-support/setup-hook: line 15: concatTo: command not found```

jcollie commented:
> Where (on which branch?) and how (any specific package to build?) can we reproduce this?

This is unfortunately in a package that isn't open source (yet), but it was on unstable. I'll see if I can reproduce it tonight in another package that is open source.
